Improves docker setup for basic usage to get started more easily

daniel-j-h opened the pull request:
Hi! :wave: Finally giving your project a try and because it's 2024 ain't nobody got time for building C++ from source, I checked out the docker images. In this changeset I make various improvements to the docker setup
1. Using the stable debian bookworm as a base image as bullseye is outdated
2. Not installing the development libraries with the tilemaker binary as they're only needed for building
3. Compiling in parallel so that the docker build doesn't take 10+ minutes to build
4. Adding a default command of `--help` so that if folks do `docker run` they get to see the tilemaker help
5. Updated the docker readme and added a usage section in the default readme to use docker by default

I would highly recommend defaulting to telling users to just use the docker image and not having to compile tilemaker from source. There are so many compilation problems caused by dependencies to install, compiler and linker versions, and so on.

Just from a first look
- https://github.com/systemed/tilemaker/issues/702
- https://github.com/systemed/tilemaker/issues/694
- https://github.com/systemed/tilemaker/issues/444

If we could dumb down the getting started instructions to

    docker run .. tilemaker input.pbf --output output.pbf

that would make it easier for users and hopefully less likely for us to get support tickets.

systemed commented:
This is great - thank you!

I know practically nothing about docker - the support that we have so far is entirely down to other contributors. So I'm very happy to merge this but I may be leaning on you and others for support in the future.

> it's 2024 ain't nobody got time for building C++ from source

Call me old-school 😉 (but also it's seriously fast on Apple silicon...).

daniel-j-h commented:
Got it, I believe it's fair to say that Docker will not go away so can recommend looking into it.

In the context of native projects like tilemaker the benefit I see
- on the dev side is being able to compile in a self-contained development environment with a specific compiler against one set of libraries, so you don't have to support the combination of gcc, clang x library versions but e.g. whatever is in Debian LTS
- on the user side the benefit is you don't have to compile anything at all but can just run the pre-built docker image

If you want to give this PR a try locally

```bash
docker build -t tilemaker .
docker run -it --rm tilemaker
```

I just rebased against master and added the tilemaker binary to PATH so https://github.com/systemed/tilemaker/pull/465 can be closed, too.

Like a said in https://github.com/systemed/tilemaker/pull/713#issuecomment-2086138281 in the future a second iteration would be publishing not just a single tilemaker docker image but tagging them e.g. by date or version or commit hash. This can be a follow-up pull request.

Happy to get this in?

leonardehrenfried commented:
Now that this one is also merged, would it be possible to take a look at https://github.com/systemed/tilemaker/pull/685 ?
